### ISSUE TITLE
Wrong receive/send order issue, and packet merge on fast SendToXXX() call.

### ISSUE THREAD

**Issue by GarlicDipping** [Eastrall/Sylver.Network#5]
Hi, first of all thanks for the great library! I love the design and simplicity of this project. 

But during playing around with this library, I've encountered two issues and wanted to share it. (With some suggestions.)

Both of the issues occured when using NetServer.SendToXXX() function **multiple times** at the **(almost)same time.**

To reproduce, Just modify 3 files from [Sylver.Chat](https://github.com/Eastrall/Sylver.Network/tree/master/samples/Chat) sample project : 

(modified code pastebin links)

1. [ChatServer.cs](https://pastebin.com/izVkxWUp) 
2. [ChatClient.cs](https://pastebin.com/svtHwdpR) 
3. [ChatPacketType.cs](https://pastebin.com/m0RG6rKb) 

Results may differ per every client run, but you'll get the point.
Problem is : 

1. Messages are not received as server sent order and
2. Some messages are dropped.

Issue 1 can cause problem when send order is important. I've poked around and found that [NetMessageDispatcher.DispatchMessage()](https://github.com/Eastrall/Sylver.Network/blob/20418d22df99ced02ad55fe0b887bae293d380a3/src/Sylver.Network/Infrastructure/NetMessageDispatcher.cs#L26) uses Task.Run(), which seems like order issue is from typical thread problem. If this is not your intended behaviour(And I believe so because packet order is one of the most important feature from TCP connection), I suggest that 

1. Either remove Task.Run() or 
2. Use similar queue-processing pattern like [NetSender.cs](https://github.com/Eastrall/Sylver.Network/blob/master/src/Sylver.Network/Infrastructure/NetSender.cs) uses.

Issue 2 is about dropped packet. Actually, packets are not dropped(cause it's TCP), they are just not properly handled in [NetPacketParser.ParseIncomingData()](https://github.com/Eastrall/Sylver.Network/blob/8ae86be587e52b7bc9f58a7c99840e2b21a0a9aa/src/Sylver.Network/Data/Internal/NetPacketParser.cs#L30). Sometimes Socket.SendAsync() automatically **merges** tiny packets and ParseIncomingData() needs some bug fix for this. What about let ParseIncomingData() return false when merged bytes are not all done, and call [DispatchMessage()](https://github.com/Eastrall/Sylver.Network/blob/8ae86be587e52b7bc9f58a7c99840e2b21a0a9aa/src/Sylver.Network/Infrastructure/NetReceiver.cs#L72) several times until every merged bytes are processed properly?

So that's for it! Thanks for reading, and let me know if there's anything I was wrong or misunderstood!

**Comment by Eastrall** [comment]
Hi, thanks for using `Sylver.Network` and for the feedback!

You are correct about your two points. A good approach as you suggested for the 1st issue would be to implement a receive queue and for each received message start a new task using the .NET task factory. Let's work on that.

The second issue is a bit tricky. All `NetPackets` are formatted with an header that contains the packet length. The parsing algorithm is based on this length to split the packets correctly (framing). Maybe it's an issue with the buffer pool, which doesn't have space anymore and some data are loss; or it might be also related to how we handle packets (issue 1).

I'm going to investigate on the second issue, and work on the first one to optimize the receive process.

Thanks again for the feedback!

**Comment by Eastrall** [comment]
I've figured out the issue for the problem 2, and you were correct. There was an issue with my algorithm when it comes to process multiple packets. Only the first one was parsed correctly and the rest was basically, broken ...

I made a fix and also implemented the producer-consumer pattern for the received messages. Pull requests is coming soon.

Once all the tests are OK, I'll publish a new version on nuget.

Thanks again for the feedback that allow us to improve the library!

**Comment by GarlicDipping** [comment]
Hi, Thanks for the quick fix! 

But... sadly there are still some problems not fixed yet. 😢 

1. First, packet order issue still exists. And the root cause is same - Task's internal threading order problem. 

As [NetReceiver.Start()](https://github.com/Eastrall/Sylver.Network/blob/87af3456282bd57e778a07d75b309da7f78037a7/src/Sylver.Network/Infrastructure/NetReceiver.cs#L43) method creating a tasked while(true) loop now, I think [NetReceiver.ProcessReceivedMessages()](https://github.com/Eastrall/Sylver.Network/blob/87af3456282bd57e778a07d75b309da7f78037a7/src/Sylver.Network/Infrastructure/NetReceiver.cs#L203) method's Task.Factory.StartNew() call is not necessary. Just remove that line and everything seems to work well. (...at least in my test 😄)

2. Sometimes, NetClient.HandleMessage() is called **before** NetClient.OnConnected().
I think it's natural for client to process any other messages from server **after** OnConnected() call. Calling OnConnected() before _receiver.Start(this) in [NetClient.OnClientConnected()](https://github.com/Eastrall/Sylver.Network/blob/87af3456282bd57e778a07d75b309da7f78037a7/src/Sylver.Network/Client/NetClient.cs#L137) will be enough for fix. 

Anyway, that packet merge issue seems to disappeared now. Great! 👍 

I'll make a pull req for the fix 1 & 2. Please check those out, and let me know if there are any mistakes or misunderstandings I have.

Thanks for your attention! Have a nice day!

**Comment by Eastrall** [comment]
The problem is, if you remove the `Task.Factory.StartNew()` method call, the receive queue will have to wait the end of the `HandleMessage(packet)` method. I don't know for sure, but this might cause performance issues. Is there a specific requirement to receive the `NetPacket`s in the correct order? Do you depend on previously sent packets?

Also, just realized that the receiver task shouldn't be created in the `NetReceiver.Start()`, but in the constructor instead. Each time a new client connects, the `Start()` method is called passing the newly connected client. So, in my current changes, each time a new client connects, it creates a new task... Which is, obviously, not a good thing in terms of performances. 😄

**Comment by GarlicDipping** [comment]
Oh, I see what you mean. 

> Is there a specific requirement to receive the NetPackets in the correct order?

No, For now I'm just learning how newtork works in c#, and was curious that many articles about TCP connections said _'Packet order is guaranteed in TCP connection, blah blah'_ but in this framework sometimes it does not. 

I just thought It'll be a huge headache to deal with _'PlayerDead packet arrived first then HealPlayer arrived next! What's going on!?'_ kind of problem, and guaranteeing packet order will remove whole a lot of these problems. But I think implementing these kind of determinism is something out of topic and now it's just some kind of design choice problem rather than bugfixes. (Which I thought was a bug 😄)

**Comment by Eastrall** [comment]
I think what the TCP documentation says about packets are that the packets you send, will always arrive in the same order, like, you send a packet with 1 `byte`, 1 `long` and 2 `int`s, it's guarantied that they will arrive in the same order and not 1 `long`, 1 `int`, 1 `byte`, 1 `int`.

When we take a message from the received message queue, the message have already been receive and parsed, but not yet handled. This is why, foreach message in the message queue, we create a new packet and make the correct client handle it.

Now, it's true, I don't know if the packets should be handled one after another; which means each clients have to wait for others to finish if packet is being processed. In my opinion, this would be nice for maybe 50 connections, but if we talk for example for a MMORPG that can handle more than 100.000 simultaneous connections, this will be very bad in terms of performances.

**Comment by GarlicDipping** [comment]
I agree with you that in a large-scaled application like MMORPG, It's inevitable to process message queue as fast as possible. (Otherwise message queue will be filled up very quickly and messages should wait until their turn comes, as you said.)

But in my opinion, if logics inside HandleMessage() takes too long... then it means that user's business logic needs optimization - which means it's up to user's choice to use Task.Factory.StartNew or Threading or something else, rather than network framework preoptimize it for user.

(*And, I just realized that... the producer-consumer pattern in NetReceiver.cs was implemented to resolve packet order thing in the first place, right? 😅 Otherwise we can just stick to Task.Run() way...)

**Comment by Eastrall** [comment]
I also agree with you that if `HandleMessage()` takes too much time to process, it needs an otpimization. But in some cases, for example when you query a database, you can have a timeout expcetion and then the action will not be handled in a reasonable time.

The producer consumer pattern helps decoupeling the parsing process from the handling. Once the socket data stream is parsed and split into Sylver packets, they are queue to be handled for the given client. I think we can keep it that way, but we will need to call `Task.Factory.StartNew()` to handle the message asynchronously; without await for every clients. 😄

**Comment by Eastrall** [comment]
I have removed the receive queue from the receiver and made it process imediatly the incoming packets on the same thread. It's basically what you have done in your PR, but without the receive queue.

Also, I have included your change on the `NetClient` class, where the `OnConnected()` was called after the `receiver.Start()` method.

Finally, I'm going to work on error handling, because I saw in some parts of the code that some error cases are not handled at all... There will be a version 1.2.0 soon enough with this new feature and fixes.
